### DISCUSSION TITLE
How to save Authenticator?

### DISCUSSION THREAD

**Discussion by epikmartin** [webauthn4j/webauthn4j#447] [Q&A] [answered]
Hi, 

first of all: thank you for this very useful framework!

We implemented WebAuthn registration and login based on this framework and my question is about saving the Authenticator.
The simplest way to save the Authenticator in a Database is to serialize it. This is what we do right now and it works fine. 
But what will happen, if one day in a new Version of the framework the Authenticator class changes? There is no explicit serialVersionUID, so Java will generate one. If it changes, we can not deserialize our previously saved AuthenticatorS which will cause a lot of trouble ;-)

What is the best way to save a Authenticator object in a DB?

Thank you and best regards!
Martin

**Comment by ynojima** [comment]
Hi, Thank you for asking.
`Authenticator` is an interface and you need to implement a concrete class and its serializer and deserializer by yourself.
For types used as `Authenticator`'s member, there are [converters](https://github.com/webauthn4j/webauthn4j/tree/master/webauthn4j-core/src/main/java/com/webauthn4j/converter) that can be used for serialization and deserialization.

**Reply by epikmartin** [reply]
Hello, 

thanke you for your quick answer!
We use com.webauthn4j.authenticator.AuthenticatorImpl
This works so far and we do not see any reason to implement our own Authenticator.

As I understand your answer, you suggest to use the provided converters to serialize the objects? This sounds reasonable and we will try to do so.
My main concern was to make sure that all serialized objects are compatible to future versions of the framework. So I assume that this is ensured by keeping the converters compatible, right?

Thanks again and have a nice day!
Martin

**Reply by santhoshit8** [reply]
Hi all,

I am also using default com.webauthn4j.authenticator.AuthenticatorImpl .

**For  storing Authenticator**

`Authenticator authenticator =
        new AuthenticatorImpl( // You may create your own Authenticator implementation to save friendly authenticator name
                registrationData.getAttestationObject().getAuthenticatorData().getAttestedCredentialData(),
                registrationData.getAttestationObject().getAttestationStatement(),
                registrationData.getAttestationObject().getAuthenticatorData().getSignCount()
        );`

`AttestedCredentialDataConverter.convert(AttestedCredentialData)`  - this convert whole AttestedCredentialData object into binary data and whole binary data into  AttestedCredentialData object . is storing whole object as binary in db good way?

1) what is best way to store AttestedCredentialData or  AuthenticatorImpl object in db using class provided in library or should i use custom serialization and deserialization?


**For loading Authenticator**

`Authenticator authenticator = new AuthenticatorImpl(AttestedCredentialData, null, counter);`

2) is passing null to getAttestationStatement good way during authentication verification.

**Reply by ynojima** [reply]
> AttestedCredentialDataConverter.convert(AttestedCredentialData) - this convert whole AttestedCredentialData object into binary data and whole binary data into AttestedCredentialData object . is storing whole object as binary in db good way?

When you store `AttestedCredentialData` in DB, it is better to breake it into 3 columns(aaguid, credentialId, coseKey) because credentialId is used as a key to lookup the `Authenticator` data.

> what is best way to store AttestedCredentialData or AuthenticatorImpl object in db using class provided in library or should i use custom serialization and deserialization?

If you are looking for JPA example, `AuthenticatorEntity` in webauthn4j-spring-security sample SPA could be a good example.
https://github.com/webauthn4j/webauthn4j-spring-security/blob/0.7.6.RELEASE/samples/spa/src/main/java/com/webauthn4j/springframework/security/webauthn/sample/domain/entity/AuthenticatorEntity.java

> is passing null to getAttestationStatement good way during authentication verification?

As of today, attestation statement is not used during authentication, and `getAttestationStatement` is `nullable`. You may pass `null`.
https://github.com/webauthn4j/webauthn4j/blob/0.20.3.RELEASE/webauthn4j-core/src/main/java/com/webauthn4j/authenticator/CoreAuthenticator.java#L45-L47

**Reply by santhoshit8** [reply]
thanks for info @ynojima

**Reply by hetin62** [reply]
Hello @ynojima 

Can you provide thought on how to store coseKey? For example, if we change library in future, is there any problem arise? In short, how to store coseKey in library independent way?

**Reply by ynojima** [reply]
Interesting question. I think saving the `credentialPublicKey` (`coseKey`) in COSE key format is the best.
When the `credentialPublicKey` (`coseKey`) is loaded from a security key, it is encoded in COSE key format. (Ref. CBOR Object Signing and Encryption (COSE): https://datatracker.ietf.org/doc/html/rfc9052)
So, if you save it in COSE key format, no information will be lost from format conversion (as long as WebAuthn4J's COSEKey classes supports), and other libraries may have a way to load the public key from the COSE key format `credentialPublicKey`. Even if other libraries don't support it, the Jackson CBOR deserializers in WebAuthn4J support standalone use, they can be used to parse the COSE key format `credentialPublicKey into Java's standard `PublicKey` implementation.

**Comment by ynojima** [comment, marked as answer]
> My main concern was to make sure that all serialized objects are compatible to future versions of the framework. So I assume that this is ensured by keeping the converters compatible, right?

Yes, we'll keep the types used in the `Authenticator` class backward compatible. If you find incompatibility in the future, please let us know. We'll try to provide a fix or workaround.

**Reply by epikmartin** [reply]
Thank you very much!
